### PR TITLE
Make the default value of E57_RELEASE_LTO dependent on E57_BUILD_SHARED

### PR DESCRIPTION
**Pull request by SunBlack**
By default [`BUILD_SHARED_LIBS`](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) is not defined, therefore the default of `E57_BUILD_SHARED` is also `OFF`. Therefore this lib will be build as static library by default. This leads by default to linker issues later:

```
/usr/bin/ld: /home/sunblack/dev/dependencies/libE57Format/lib/libE57Format.a(BlobNode.cpp.o): plugin needed to handle lto object
```

As the comment about `E57_RELEASE_LTO` says, this option should be OFF when building a static library. Therefore the default value is now adjusted accordingly. As changing `E57_BUILD_SHARED` later will not change the default value of `E57_RELEASE_LTO` I have added a warning so that there is a hint that the combination may cause issues.

GDAL has IPO disabled by default:
https://github.com/OSGeo/gdal/blob/305b23f5a90bf6d86c7fb895dfc745321def0aab/gdal.cmake#L27

Qt seems to also [not set this value](https://github.com/search?q=repo%3Aqt%2Fqtbase%20INTERPROCEDURAL_OPTIMIZATION&type=code), but in case `CMAKE_INTERPROCEDURAL_OPTIMIZATION` is set they enforce it to `OFF` for MSVC 3rd party:
https://github.com/qt/qtbase/blob/dev/cmake/Qt3rdPartyLibraryHelpers.cmake#L334-L338
Reason see [here](https://github.com/qt/qtbase/commit/0d708a9aad8fa7f44e94f5bc9093a6c93d6140b).

So basically I would argue that `E57_RELEASE_LTO` should be always `OFF` or removed and in case someone needs this, he can build it by passing `CMAKE_INTERPROCEDURAL_OPTIMIZATION` (in case canon has an issue with it, they can add it to their build script as it seems it was not necessary for vcpkg when checking the port scripts from before libE57Format 3.1.0 release)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -113,7 +113,13 @@ option( E57_WRITE_CRAZY_PACKET_MODE "Compile library to enable reader-stressing
 # CMake forces "thin" LTO (see https://gitlab.kitware.com/cmake/cmake/-/issues/23136)
 # which is a problem if compiling statically for distribution (e.g. in a package manager).
 # Generally you will only want to turn this off for distributing static release builds.
-option( E57_RELEASE_LTO "Compile release library with link-time optimization" ON )
+option( E57_RELEASE_LTO "Compile release library with link-time optimization" ${E57_BUILD_SHARED} )
+
+if ( E57_RELEASE_LTO AND NOT E57_BUILD_SHARED )
+    message( WARNING "E57_RELEASE_LTO is set to ON, but ${PROJECT_NAME} will be build as static library. This may lead to linker issues!" )
+elseif ( NOT E57_RELEASE_LTO AND E57_BUILD_SHARED )
+	message( WARNING "E57_RELEASE_LTO is set to OFF, but ${PROJECT_NAME} will be build as dynamic library. This may lead issues if compiling statically for distribution!" )
```
Comment (asmaloney) [inline CMakeLists.txt:121]: Having it OFF for a dynamic lib isn't a problem. If they are compiling a dynamic lib, then they are are not "compiling statically for distribution".

Maybe I'm misunderstanding what you're trying to warn about here?

Comment (SunBlack) [reply CMakeLists.txt:121]: I have only copied text from the comment, as it only says that there are problems, but nothing more. From #254 I also couldn't figure out why the option really has to be set as a default and couldn't just be added from outside if needed.

**Comment by SunBlack** [comment]
> If you're using a static lib in your own code (i.e. including this library as a static lib) it's fine.

Sadly not. See the linker issue in the initial post of me.

File: CMakeLists.txt (hunk 1 above)
Comment (asmaloney) [reply CMakeLists.txt:121]: If you're trying to package the lib for distribution, you can just turn it off by setting `E57_RELEASE_LTO` to `OFF`.

**Comment by asmaloney** [comment]
I use it as a static lib on macOS and Windows (MSYS2) without issues so this sounds like maybe a compiler difference.

What OS and compiler are you using?

**Comment by SunBlack** [comment]
> I use it as a static lib on macOS and Windows (MSYS2) without issues so this sounds like maybe a compiler difference.
> 
> What OS and compiler are you using?

On MSVC everything is fine with 3.2.0. On our CI we are using a docker image based on `ubuntu:22.04` (my local VM is also Ubuntu 22.04 and there I have the same issue). Compiler is clang (19) and gcc (11).

**Comment by asmaloney** [comment]
I've never seen "plugin needed to handle lto object" - sounds like something isn't installed? Are you getting the same message from both clang and gcc on Ubuntu?

Any more info about it when you search for it?

**Comment by SunBlack** [comment]
> I've never seen "plugin needed to handle lto object" - sounds like something isn't installed? Are you getting the same message from both clang and gcc on Ubuntu?

Tried again. On GCC I'm getting:
```
lto1: fatal error: bytecode stream in file ‘/home/sunblack/dev/dependencies/libE57Format/lib/libE57Format.a’ generated with LTO version 11.3 instead of the expected 12.0
```

So basically LTO seems to break the use case, to build a dependency with the default OS compiler and then compile it with a different compiler. Since we test our project on our CI with different GCC versions (standard OS compiler when we build a release so that we don't have to install another standard lib at the customer's site) and the latest compiler for more compiler warnings, you have to build multiple versions of libE57 (if you have LTO enabled) so that you can build with multiple compilers at the same time.

In principle, I have the feeling that you really don't want LTO as default, unless you build a release where you want to have a better optimization again - and in that case you can also activate it manually.

> Any more info about it when you search for it?

[Here](https://github.com/awslabs/aws-lambda-cpp/issues/128#issuecomment-1237497849) for example

**Comment by asmaloney** [comment]
OK thanks. I'll consider it.

In the meantime, you should be able to set `E57_RELEASE_LTO` to `OFF`.

**Comment by StefanBruens** [comment]
@SunBlack Your setup is just combining the worst of all worlds.

When distributing the library (even for internal use), you should use a shared library. This provides the benefit of better optimization of calls inside the library (i.e. one libE57format function calling another), and better diagnostics (as the compilier can "see" both caller and callee, it can analyze the code in more depth).

For even more optimization and better diagnostics, you should include libE57format as e.g. a submodule, and compile everything with the same compiler. Then you won't have any problems with different versions.  Or you build it separately multiple times, but use the same compiler+linker for all steps.

Your use case is IMHO an outlier. The two typical cases are either static linking, building everything combined; or building a shared library. Only if you want to distribute the static library for reuse by others (where you have no control or knowledge about the used toolchain) a static library with regular object files is useful.

**Comment by nh2** [comment]
I also hit this problem when building my app with the [nixpkgs](https://github.com/NixOS/nixpkgs/) package distribution.

When upgrading from NixOS 24.05 to 25.05, which has the upgrade from 2.2.0 to 3.1.1 via [my commit](https://github.com/NixOS/nixpkgs/commit/a8e7538fcbd223f4e521796d680d6520f908d799), `E57_RELEASE_LTO` got enabled.

This makes that in my downstream app built with nix that discovers `libE57Format` via Meson and thus passes `/nix/store/bl0qh9rx56w6vrha03d4ch2i7qgbi2v0-libe57format-3.2.0/lib/libE57Format.a` to `g++` for linking, I get linker error:

```
/nix/store/...-binutils-2.44/bin/ld: /root/benaco-nixos-25.05/serkezet/cpp/build-incremental/../src/E57.cpp:91:(.text+0x155):
  undefined reference to `e57::ImageFile::root() const'
```

When Linux distributions like NixOS build the `.a` file, should they then pass `-DE57_RELEASE_LTO=OFF` to CMake?
